File: packages/core/src/programmers/llm/LlmControllerProgrammer.ts
import { ILlmSchema } from "@typia/interface";
import ts from "typescript";

import { ITypiaContext } from "../../context/ITypiaContext";
import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";
import { LlmApplicationProgrammer } from "./LlmApplicationProgrammer";

export namespace LlmControllerProgrammer {
  export interface IWriteProps {
    context: ITypiaContext;
    modulo: ts.LeftHandSideExpression;
    metadata: MetadataSchema;
    config?: Partial<
      ILlmSchema.IConfig & {
        equals: boolean;
      }
    >;
    className?: string;
    node: ts.TypeNode;
    nameArgument: ts.Expression;
    executeArgument: ts.Expression;
    configArgument?: ts.Expression;
  }

  export const write = (props: IWriteProps): ts.Expression => {
    const typeNode: ts.ImportTypeNode = props.context.importer.type({
      file: "typia",
      name: "ILlmController",
      arguments: [props.node],
    });
    const value: ts.ObjectLiteralExpression =
      ts.factory.createObjectLiteralExpression(
        [
          ts.factory.createPropertyAssignment(
            "protocol",
            ts.factory.createStringLiteral("class"),
          ),
          ts.factory.createPropertyAssignment("name", props.nameArgument),
          ts.factory.createPropertyAssignment("execute", props.executeArgument),
          ts.factory.createPropertyAssignment(
            "application",
            LlmApplicationProgrammer.write({
              context: props.context,
              modulo: props.modulo,
              metadata: props.metadata,
              config: props.config,
              name: props.className,
              configArgument: props.configArgument,
            }),
          ),
        ],
        true,
      );
    return ts.factory.createAsExpression(
      ts.factory.createSatisfiesExpression(value, typeNode),
      typeNode,
    );
  };
}

File: packages/typia/native/core/programmers/llm/LlmControllerProgrammer.go
package llm

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type llmControllerProgrammerNamespace struct{}

var LlmControllerProgrammer = llmControllerProgrammerNamespace{}

type LlmControllerProgrammer_IWriteProps struct {
  Context         nativecontext.ITypiaContext
  Modulo          *shimast.Node
  Metadata        *schemametadata.MetadataSchema
  Config          map[string]any
  ClassName       *string
  Node            *shimast.Node
  NameArgument    *shimast.Node
  ExecuteArgument *shimast.Node
  ConfigArgument  *shimast.Node
}

var llmControllerProgrammer_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (llmControllerProgrammerNamespace) Write(props LlmControllerProgrammer_IWriteProps) *shimast.Node {
  f := nativecontext.EmitFactoryOf(llmControllerProgrammer_factory, props.Context.Emit)
  typeNode := llmProgrammer_import_type(props.Context, nativecontext.ImportProgrammer_TypeProps{
    File:      "typia",
    Name:      "ILlmController",
    Arguments: []*shimast.TypeNode{props.Node},
  })
  value := f.NewObjectLiteralExpression(f.NewNodeList([]*shimast.Node{
    f.NewPropertyAssignment(nil, f.NewStringLiteral("protocol", shimast.TokenFlagsNone), nil, nil, f.NewStringLiteral("class", shimast.TokenFlagsNone)),
    f.NewPropertyAssignment(nil, f.NewStringLiteral("name", shimast.TokenFlagsNone), nil, nil, props.NameArgument),
    f.NewPropertyAssignment(nil, f.NewStringLiteral("execute", shimast.TokenFlagsNone), nil, nil, props.ExecuteArgument),
    f.NewPropertyAssignment(nil, f.NewStringLiteral("application", shimast.TokenFlagsNone), nil, nil, LlmApplicationProgrammer.Write(LlmApplicationProgrammer_IWriteProps{
      Context:        props.Context,
      Modulo:         props.Modulo,
      Metadata:       props.Metadata,
      Config:         props.Config,
      Name:           props.ClassName,
      ConfigArgument: props.ConfigArgument,
    })),
  }), true)
  return f.NewAsExpression(
    f.NewSatisfiesExpression(value, typeNode),
    typeNode,
  )
}
